README.md
# 4850lab7
Lab 7 XML, Lab 8 XML Processing, Lab 9 XML Schema

## Conventions

Allman style indentation
UCfirst classes and files
lowercase with underscores for methods and properties

## Team Members
- Chris Leclair: XCVG
- Richard Gosse: rwgosse


changelog.txt
Newest changes at the top
Format: MM/DD/YYYY HH:MM - comment  

03/31/2016 14:35 - stylesheet for timetable with timeslots on top, added a thursday entry to day facet
03/31/2016 14:16 - stylesheet for timetable with days on top, some css
03/31/2016 13:00 - start of lab10, initial setup of stylesheet
03/24/2016 16:41 - added schema validation functionality to site
03/24/2016 16:17 - added days facet to schema and clarified the schema
03/24/2016 15:26 - fixed issues with xsd (course facet) and fixed some formatting
03/24/2016 15:05 - created course facet elements in xsd
03/24/2016 14:29 - added room datatype with regex match
03/24/2016 14:15 - setup initial datatypes for xml schema
03/21/2016 13:27 - added display of previous valid result
03/21/2016 13:01 - added common result in xml for bingo condition & bingo code to controller & view
03/11/2016 16:01 - added dropdown list to Welcome that shows all bookings matching day and timeslot
03/10/2016 16:26 - added display of timetables in Welcome controller
03/10/2016 15:37 - added null handling to Timetable.php
03/10/2016 15:13 - finished third facet
03/10/2016 14:50 - finished second facet
03/10/2016 14:32 - finished first facet
03/10/2016 12:51 - added CodeIgniter starter
03/03/2016 20:10 - added full test data, fixed multiplicity in dtd
03/03/2016 14:56 - modified DTD for courses, added test data to XML
03/03/2016 14:41 - modified DTD for timeslots view, added test data to XML
03/03/2016 14:31 - DTD, setup elements and attribs for days
03/03/2016 13:52 - File and folder setup
03/03/2016 13:41 - Initial setup


application/views/welcome.php
<?php
defined('BASEPATH') OR exit('No direct script access allowed');
?><!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <title>{title}</title>
    </head>

    <body>

    <div id="container">
        
        <div id="bookings-validation">
            
            <h1>Validation Status</h1>
            
            <p>{validation-status}</p>
            
            {validation-messages}
            <p>{validation-message}</p>
            {/validation-messages}
            
        </div>

        <div id="bookings-search">
            
            <h1>Bookings Search</h1>            
            
            <form name="search-form" action='welcome/search' id="dropdown-days">
                <select name="select-days">
                    {select-days}
                        <option value="{key}">{value}</option>
                    {/select-days}
                </select>
                <select name="select-timeslots">
                    {select-timeslots}
                        <option value="{key}">{value}</option>
                    {/select-timeslots}
                </select>
                <input type='submit' value='Submit'>
            </form>
            
        </div>
        
        <div id="bookings-lists">
            <h1>Bookings by Day</h1>
            {bookings-days}    

            <h1>Bookings by Timeslot</h1>
            {bookings-timeslots}

            <h1>Bookings by Course</h1>
            {bookings-courses}
        </div>
            
        <div id="bookings-result">
            <h1>Previous Search Result</h1>
            {bookings-result}
        </div>

    </div>

    </body>
</html>

application/config/autoload.php
<?php
defined('BASEPATH') OR exit('No direct script access allowed');

/*
| -------------------------------------------------------------------
| AUTO-LOADER
| -------------------------------------------------------------------
| This file specifies which systems should be loaded by default.
|
| In order to keep the framework as light-weight as possible only the
| absolute minimal resources are loaded by default. For example,
| the database is not connected to automatically since no assumption
| is made regarding whether you intend to use it.  This file lets
| you globally define which systems you would like loaded with every
| request.
|
| -------------------------------------------------------------------
| Instructions
| -------------------------------------------------------------------
|
| These are the things you can load automatically:
|
| 1. Packages
| 2. Libraries
| 3. Drivers
| 4. Helper files
| 5. Custom config files
| 6. Language files
| 7. Models
|
*/

/*
| -------------------------------------------------------------------
|  Auto-load Packages
| -------------------------------------------------------------------
| Prototype:
|
|  $autoload['packages'] = array(APPPATH.'third_party', '/usr/local/shared');
|
*/
$autoload['packages'] = array();

/*
| -------------------------------------------------------------------
|  Auto-load Libraries
| -------------------------------------------------------------------
| These are the classes located in system/libraries/ or your
| application/libraries/ directory, with the addition of the
| 'database' library, which is somewhat of a special case.
|
| Prototype:
|
|	$autoload['libraries'] = array('database', 'email', 'session');
|
| You can also supply an alternative library name to be assigned
| in the controller:
|
|	$autoload['libraries'] = array('user_agent' => 'ua');
*/
$autoload['libraries'] = array('parser','session');

/*
| -------------------------------------------------------------------
|  Auto-load Drivers
| -------------------------------------------------------------------
| These classes are located in system/libraries/ or in your
| application/libraries/ directory, but are also placed inside their
| own subdirectory and they extend the CI_Driver_Library class. They
| offer multiple interchangeable driver options.
|
| Prototype:
|
|	$autoload['drivers'] = array('cache');
*/
$autoload['drivers'] = array();

/*
| -------------------------------------------------------------------
|  Auto-load Helper Files
| -------------------------------------------------------------------
| Prototype:
|
|	$autoload['helper'] = array('url', 'file');
*/
$autoload['helper'] = array();

/*
| -------------------------------------------------------------------
|  Auto-load Config files
| -------------------------------------------------------------------
| Prototype:
|
|	$autoload['config'] = array('config1', 'config2');
|
| NOTE: This item is intended for use ONLY if you have created custom
| config files.  Otherwise, leave it blank.
|
*/
$autoload['config'] = array();

/*
| -------------------------------------------------------------------
|  Auto-load Language files
| -------------------------------------------------------------------
| Prototype:
|
|	$autoload['language'] = array('lang1', 'lang2');
|
| NOTE: Do not include the "_lang" part of your file.  For example
| "codeigniter_lang.php" would be referenced as array('codeigniter');
|
*/
$autoload['language'] = array();

/*
| -------------------------------------------------------------------
|  Auto-load Models
| -------------------------------------------------------------------
| Prototype:
|
|	$autoload['model'] = array('first_model', 'second_model');
|
| You can also supply an alternative model name to be assigned
| in the controller:
|
|	$autoload['model'] = array('first_model' => 'first');
*/
$autoload['model'] = array();


application/config/constants.php
<?php
defined('BASEPATH') OR exit('No direct script access allowed');

/*
|--------------------------------------------------------------------------
| Display Debug backtrace
|--------------------------------------------------------------------------
|
| If set to TRUE, a backtrace will be displayed along with php errors. If
| error_reporting is disabled, the backtrace will not display, regardless
| of this setting
|
*/
defined('SHOW_DEBUG_BACKTRACE') OR define('SHOW_DEBUG_BACKTRACE', TRUE);

/*
|--------------------------------------------------------------------------
| File and Directory Modes
|--------------------------------------------------------------------------
|
| These prefs are used when checking and setting modes when working
| with the file system.  The defaults are fine on servers with proper
| security, but you may wish (or even need) to change the values in
| certain environments (Apache running a separate process for each
| user, PHP under CGI with Apache suEXEC, etc.).  Octal values should
| always be used to set the mode correctly.
|
*/
defined('FILE_READ_MODE')  OR define('FILE_READ_MODE', 0644);
defined('FILE_WRITE_MODE') OR define('FILE_WRITE_MODE', 0666);
defined('DIR_READ_MODE')   OR define('DIR_READ_MODE', 0755);
defined('DIR_WRITE_MODE')  OR define('DIR_WRITE_MODE', 0755);


//location of data folders
define('DATAPATH', './data/');
define('XMLSUFFIX', '.xml');


/*
|--------------------------------------------------------------------------
| File Stream Modes
|--------------------------------------------------------------------------
|
| These modes are used when working with fopen()/popen()
|
*/
defined('FOPEN_READ')                           OR define('FOPEN_READ', 'rb');
defined('FOPEN_READ_WRITE')                     OR define('FOPEN_READ_WRITE', 'r+b');
defined('FOPEN_WRITE_CREATE_DESTRUCTIVE')       OR define('FOPEN_WRITE_CREATE_DESTRUCTIVE', 'wb'); // truncates existing file data, use with care
defined('FOPEN_READ_WRITE_CREATE_DESCTRUCTIVE') OR define('FOPEN_READ_WRITE_CREATE_DESTRUCTIVE', 'w+b'); // truncates existing file data, use with care
defined('FOPEN_WRITE_CREATE')                   OR define('FOPEN_WRITE_CREATE', 'ab');
defined('FOPEN_READ_WRITE_CREATE')              OR define('FOPEN_READ_WRITE_CREATE', 'a+b');
defined('FOPEN_WRITE_CREATE_STRICT')            OR define('FOPEN_WRITE_CREATE_STRICT', 'xb');
defined('FOPEN_READ_WRITE_CREATE_STRICT')       OR define('FOPEN_READ_WRITE_CREATE_STRICT', 'x+b');

/*
|--------------------------------------------------------------------------
| Exit Status Codes
|--------------------------------------------------------------------------
|
| Used to indicate the conditions under which the script is exit()ing.
| While there is no universal standard for error codes, there are some
| broad conventions.  Three such conventions are mentioned below, for
| those who wish to make use of them.  The CodeIgniter defaults were
| chosen for the least overlap with these conventions, while still
| leaving room for others to be defined in future versions and user
| applications.
|
| The three main conventions used for determining exit status codes
| are as follows:
|
|    Standard C/C++ Library (stdlibc):
|       http://www.gnu.org/software/libc/manual/html_node/Exit-Status.html
|       (This link also contains other GNU-specific conventions)
|    BSD sysexits.h:
|       http://www.gsp.com/cgi-bin/man.cgi?section=3&topic=sysexits
|    Bash scripting:
|       http://tldp.org/LDP/abs/html/exitcodes.html
|
*/
defined('EXIT_SUCCESS')        OR define('EXIT_SUCCESS', 0); // no errors
defined('EXIT_ERROR')          OR define('EXIT_ERROR', 1); // generic error
defined('EXIT_CONFIG')         OR define('EXIT_CONFIG', 3); // configuration error
defined('EXIT_UNKNOWN_FILE')   OR define('EXIT_UNKNOWN_FILE', 4); // file not found
defined('EXIT_UNKNOWN_CLASS')  OR define('EXIT_UNKNOWN_CLASS', 5); // unknown class
defined('EXIT_UNKNOWN_METHOD') OR define('EXIT_UNKNOWN_METHOD', 6); // unknown class member
defined('EXIT_USER_INPUT')     OR define('EXIT_USER_INPUT', 7); // invalid user input
defined('EXIT_DATABASE')       OR define('EXIT_DATABASE', 8); // database error
defined('EXIT__AUTO_MIN')      OR define('EXIT__AUTO_MIN', 9); // lowest automatically-assigned error code
defined('EXIT__AUTO_MAX')      OR define('EXIT__AUTO_MAX', 125); // highest automatically-assigned error code


application/controllers/Welcome.php
<?php
defined('BASEPATH') OR exit('No direct script access allowed');

class Welcome extends CI_Controller {

    public function index()
    {
        //load models
        $this->load->model('timetable');
        $this->load->model('booking');
        
        //define data array and define title
        $data = array();
        $data['title'] = "Bookings";
        
        //get and display validation messages
        $data['validation-status'] = $this->timetable->get_validation_status() ? "Timetable data validated successfully against schema!" : "Errors detected in timetable document.";
        $validation_messages = array();
        foreach($this->timetable->get_validation_messages() as $message)
        {
            $validation_messages[] = array('validation-message'=>$message);
        }
        $data['validation-messages'] = $validation_messages;       
        
        //create bookings by days section
        $bookings_by_days = $this->timetable->get_bookings_by_days();
        $fragments_by_days = '';
        foreach($bookings_by_days as $booking)
        {            
            //parse view fragment and add to fragments string
            $fragments_by_days .= $this->parser->parse('single_booking',$booking, TRUE);
        }
        $data['bookings-days'] = $fragments_by_days;
        
        //create bookings by timeslots section
        $bookings_by_timeslots = $this->timetable->get_bookings_by_timeslots();
        $fragments_by_timeslots = '';
        foreach($bookings_by_timeslots as $booking)
        {            
            //parse view fragment and add to fragments string
            $fragments_by_timeslots .= $this->parser->parse('single_booking',$booking, TRUE);
        }
        $data['bookings-timeslots'] = $fragments_by_timeslots;
        
        //create bookings by courses section
        $bookings_by_courses = $this->timetable->get_bookings_by_courses();
        $fragments_by_courses = '';
        foreach($bookings_by_courses as $booking)
        {            
            //parse view fragment and add to fragments string
            $fragments_by_courses .= $this->parser->parse('single_booking',$booking, TRUE);
        }
        $data['bookings-courses'] = $fragments_by_courses;
        
        //fill booking search lists
        $list_weekdays = array();
        foreach($this->timetable->get_list_weekdays() as $weekday => $weekday_value)
        {
            $list_weekdays[] = array('key'=>$weekday,'value'=>$weekday_value);
        }
        $data['select-days'] = $list_weekdays;
        
        $list_timeslots = array();
        foreach($this->timetable->get_list_timeslots() as $timeslot => $timeslot_value)
        {
            $list_timeslots[] = array('key'=>$timeslot,'value'=>$timeslot_value);
        }
        $data['select-timeslots'] = $list_timeslots;
        
        //get last result from session data and insert
        $last_result = $this->session->userdata('last_result');
        
        if(isset($last_result))
        {
            $data['bookings-result'] = $this->parser->parse('single_booking',$last_result, TRUE);
        }
        else
        {
            $data['bookings-result'] = 'No previous valid result.';
        }
        
        //parse outer template
        $this->parser->parse('welcome', $data);
    }
    
    function search()
    { 
        //load models
        $this->load->model('timetable');
        $this->load->model('booking');
        
        //define data and results array (inner and outer)
        $data = array();       
        $results = array();
        $multi_bookings = false; // boolean for error condition of multiple booking responses, should never happen
        $bingo = false; // check for bingo condition
        
        //get day and slot from get/post data
        $day = $this->input->get_post('select-days', TRUE);
        $slot = $this->input->get_post('select-timeslots', TRUE);
        
        //create (ugly) custom page title
        $data['title'] = "Results for " . $day . ":" . $slot;
      
        //get matching bookings by day and add to results array
        foreach($this->timetable->search_bookings_by_days($day, $slot) as $booking)
        {
            if (isset($day_booking)) // check for multiple booking responses, should never happen
            {
                $multi_bookings = true;
            }
            $booking = (array)$booking; //convert to array
            $day_booking = $booking; // for bingo check
            $booking['facet'] = "By Day"; //add facet info
            $results[] = $booking;
        }
        
        //get matching bookings by timeslot and add to results array
        foreach($this->timetable->search_bookings_by_timeslots($day, $slot) as $booking)
        {
            if (isset($time_booking)) // check for multiple booking responses, should never happen
            {
                $multi_bookings = true;
            }
            $booking = (array)$booking; //convert to array
            $time_booking = $booking; // for bingo check
            $booking['facet'] = "By Timeslot"; //add facet info
            $results[] = $booking;
        }
        
        //get matching bookings by course and add to results array
        foreach($this->timetable->search_bookings_by_courses($day, $slot) as $booking)
        {
            if (isset($course_booking)) // check for multiple booking responses, should never happen
            {
                $multi_bookings = true;
            }
            $booking = (array)$booking; //convert to array
            $course_booking=$booking; //for bingo check
            $booking['facet'] = "By Courses"; //add facet info
            $results[] = $booking;
        }
        
        //check for bingo condition
        if (!$multi_bookings) 
            {
                if (isset($day_booking) && isset($time_booking) && isset($course_booking))
                {
                 if ($day_booking == $time_booking && $day_booking == $course_booking )
                    {
                      $bingo = true;
                    }
                }
        }
        
        if ($bingo) // show single common result if all 3 match
        {
            $time_booking['facet'] = "Any";
            $data['results'] = array($time_booking);
            $data['bingo'] = "BINGO!!!";
            
            //set userdata
            $this->session->set_userdata('last_result',$time_booking);
         }
        else 
        {    
            $data['bingo'] = "";
            
            //put the results in data
            $data['results'] = $results;

            //set userdata
            $this->session->set_userdata('last_result',NULL);            
        }  
        
        //parse the template
        $this->parser->parse('booking_results', $data);
    }
}


application/models/Timetable.php
<?php

/* 
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

class Timetable extends CI_model 
    {
    
    private static $list_weekdays =  array('mon'=>'Monday','tues'=>'Tuesday','weds'=>'Wednesday','thurs'=>'Thursday','fri'=>'Friday');
    private static $list_timeslots = array('0830'=>'8:30-9:20', '0930'=>'9:30-10:20', '1030'=>'10:30-11:20', '1130'=>'11:30-12:20', '1230'=>'12:30-13:20',
                                             '1330'=>'13:30-14:20', '1430'=>'14:30-15:20', '1530'=>'15:30-16:20', '1630'=>'16:30-17:20');
     
    protected $xml = null;
    protected $validation_status = FALSE;
    protected $validation_messages = array();
    protected $bookings_by_days = array();
    protected $bookings_by_timeslots = array();
    protected $bookings_by_courses = array();
    
    
    //constructor
    function __construct() 
    {

        $this->xml = simplexml_load_file(DATAPATH . 'timetable.xml');
        
        //validate and retain status
        $doc = dom_import_simplexml($this->xml)->ownerDocument;
        libxml_use_internal_errors(true);
        if($doc->schemaValidate(DATAPATH . 'timetable.xsd'))
        {
            $this->validation_status = TRUE;
        }
        else
        {
            $this->validation_status = FALSE;
            foreach(libxml_get_errors() as $error)
            {
                $this->validation_messages[] = $error->message;
            }
        }
        libxml_clear_errors();

        //first facet
        foreach ($this->xml->days->day as $day)
        {
            //a day can have more than one booking
            foreach($day->timeslot as $timeslot)
            {
                //a timeslot must have only one booking in this context
                $booking_details = array(); //create details array
                
                //get the weekday as attribute from day element
                $booking_details['day_of_week'] = $day['weekday']; 
                
                //get start and end of timeslot from attributes of timeslot
                $booking_details['timeslot_start'] = $timeslot['start-time'];
                $booking_details['timeslot_end'] = $timeslot['end-time'];
                
                //get the course element and its attributes
                $course = $timeslot[0]->course;
                $booking_details['course_program'] = $course['program'];
                $booking_details['course_code'] = $course['code'];
                
                //get the booking element, attribute, and contents
                $booking = $course[0]->booking;
                $booking_details['instructor'] = $booking['instructor'];
                $booking_details['room'] = $booking;
                                          
                //build the booking object and add to array
                $this->bookings_by_days[] = new Booking($booking_details);
                
            }
            
        }
        //second facet
        foreach ($this->xml->timeslots->timeslot as $timeslot) 
        {
            // a timeslot can exist on many days
            foreach ($timeslot->day as $day)
            {
                $booking_details = array(); 
                
                //get start and end of timeslot from attributes of timeslot
                $booking_details['timeslot_start'] = $timeslot['start-time'];
                $booking_details['timeslot_end'] = $timeslot['end-time'];
                
                //get the weekday as attribute from day element
                $booking_details['day_of_week'] = $day['weekday']; 
                
                //get the course element and its attributes
                $course = $day[0]->course;
                $booking_details['course_program'] = $course['program'];
                $booking_details['course_code'] = $course['code'];
                
                //get the booking element, attribute, and contents
                $booking = $course[0]->booking;
                $booking_details['instructor'] = $booking['instructor'];
                $booking_details['room'] = $booking;
                
                //build the booking object and add to array
                $this->bookings_by_timeslots[] = new Booking($booking_details);
            }
            
        }
        
        //third facet
        foreach ($this->xml->courses->course as $course)
        {
            // a course can be on multiple days
            foreach ($course->day as $day)
            {
                // a course could happen multiple times per day
                foreach ($day->timeslot as $timeslot)
                {
                    $booking_details = array();
                    
                    //get the course element and its attributes
                    
                    $booking_details['course_program'] = $course['program'];
                    $booking_details['course_code'] = $course['code'];
                    
                    //get the weekday as attribute from day element
                    $booking_details['day_of_week'] = $day['weekday']; 
                    
                    //get start and end of timeslot from attributes of timeslot
                    $booking_details['timeslot_start'] = $timeslot['start-time'];
                    $booking_details['timeslot_end'] = $timeslot['end-time'];
                    
                    //get the booking element, attribute, and contents
                    $booking = $timeslot[0]->booking;
                    $booking_details['instructor'] = $booking['instructor'];
                    $booking_details['room'] = $booking;
                    
                    $this->bookings_by_courses[] = new Booking($booking_details);
                    
                }                
            }
        }
        
    }
    
    //search methods
    function search_bookings_by_days($day,$slot)
    {
        //results array
        $results = array();
        
        //iterate through and add all matching entries to results array
        foreach($this->bookings_by_days as $booking)
        {
            if($booking->day_of_week === $day && $booking->timeslot_start === $slot)
            {
                $results[] = $booking;
            }
        }
        
        return $results;
    }
    
    function search_bookings_by_timeslots($day,$slot)
    {
        //results array
        $results = array();
        
        //iterate through and add all matching entries to results array
        foreach($this->bookings_by_timeslots as $booking)
        {
            if($booking->day_of_week === $day && $booking->timeslot_start === $slot)
            {
                $results[] = $booking;
            }
        }
        
        return $results;
    }
    
    function search_bookings_by_courses($day,$slot)
    {
        //results array
        $results = array();
        
        //iterate through and add all matching entries to results array
        foreach($this->bookings_by_courses as $booking)
        {
            if($booking->day_of_week === $day && $booking->timeslot_start === $slot)
            {
                $results[] = $booking;
            }
        }
        
        return $results;
    }
    
    //accessors
    
    function get_bookings_by_days()
    {
        return isset($this->bookings_by_days)? $this->bookings_by_days : null;
    }

    function get_bookings_by_timeslots()
    {
        return isset($this->bookings_by_timeslots) ? $this->bookings_by_timeslots : null;
    }
    
    function get_bookings_by_courses()
    {
        return isset($this->bookings_by_courses) ? $this->bookings_by_courses : null;
    }
    
    function get_validation_status()
    {
        return $this->validation_status;
    }
    
    function get_validation_messages()
    {
        return $this->validation_messages;
    }
    
    static function get_list_weekdays() {
        return self::$list_weekdays;
    }

    static function get_list_timeslots() {
        return self::$list_timeslots;
    }
    


  
}


class Booking extends CI_model
{
    public $course_program;
    public $course_code;
    public $day_of_week;
    public $timeslot_start;
    public $timeslot_end;
    public $instructor;
    public $room;
    
    //constructor, takes associative array
    //because of the way our XML is designed, it would be very cumbersome to handle XML structure here
    //instead, we will handle it in the timetable class and build Booking to have no knowledge of it
    function __construct($details = null) //default value to fix construct problems
    {

        $this->course_program = (String) $details['course_program'];
        $this->course_code = (String) $details['course_code'];
        $this->day_of_week = (String) $details['day_of_week'];
        $this->timeslot_start = (String) $details['timeslot_start'];
        $this->timeslot_end = (String) $details['timeslot_end'];
        $this->instructor = (String) $details['instructor'];
        $this->room = (String) $details['room'];        
    }
    
    
}